imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -418,76 +418,157 @@ def set_spin_cycle_from_spin_data(
+    spin_met_per_asc : xr.DataArray
+        An array of spin acquisition epochs in MET seconds.
+
+    Returns
+    -------
+    closest_start_acq_indices : np.ndarray
+        An array of indices representing the closest spin acquisition times for each
+        science event.
+    time_diff : np.ndarray
+        An array of time differences between each science event and all spin acquisition
+        times.
+    """
+    # Find the closest start_acq for each direct event
+    # computes the index of the closest spin_met_per_asc for each science_met_per_asc
+    # so the resulting array will be of length len(science_met_per_asc), one index per
+    # ASC, but the value of each index will be the index of the closest spin data.
+    # Find the index of the spin data that starts just before or at each science epoch
+    time_diff = science_met_per_asc[:, None] - spin_met_per_asc
+    time_diff_masked = np.where(time_diff >= 0, time_diff, np.inf)
+    closest_start_acq_indices = time_diff_masked.argmin(axis=1)
+
+    return closest_start_acq_indices, time_diff
+
+
+def find_valid_asc(
+    time_diff: np.ndarray,
+    closest_start_acq_indices: np.ndarray,
+    spin_data: xr.Dataset,
+) -> np.ndarray:
+    """
+    Find valid Aggregated Science Cycles by filtering invalid spin data.
+
+    Parameters
+    ----------
+    time_diff : np.ndarray
+        Time differences between science and spin data.
+    closest_start_acq_indices : np.ndarray
+        Indices of closest spin acquisitions.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+
+    Returns
+    -------
+    valid_mask : np.ndarray
+        Boolean mask indicating valid ASCs.
+    """
+    # Start with all ASCs as valid
+    valid_mask = np.ones(len(closest_start_acq_indices), dtype=bool)
+
+    # Filter out epochs where no valid spin data exists (all time_diff < 0)
+    no_prior_spin = np.all(time_diff < 0, axis=1)
+    if no_prior_spin.any():
+        logger.warning(f"Dropping {no_prior_spin.sum()} ASCs with no prior spin data")
+        valid_mask &= ~no_prior_spin
+
+    # Filter out epochs with invalid indices
+    invalid_indices = closest_start_acq_indices < 0
+    if invalid_indices.any():
+        logger.warning(
+            f"Dropping {invalid_indices.sum()} ASCs with invalid spin indices"
+        )
+        valid_mask &= ~invalid_indices
+
+    # Filter out ASCs with insufficient spins
+    valid_cycles = spin_data["num_completed"] == 28
+    # Only check cycles for valid indices
+    valid_indices = closest_start_acq_indices[valid_mask]
+    insufficient_spins = ~valid_cycles.values[valid_indices]
+    if insufficient_spins.any():
+        logger.warning(
+            f"Dropping {insufficient_spins.sum()} ASCs with fewer than 28 spins"
+        )
+        # Update the mask for remaining valid entries
+        temp_mask = np.copy(valid_mask)
+        temp_mask[valid_mask] &= ~insufficient_spins
+        valid_mask = temp_mask
```

Comment:
Rather than this subsetting of valid indices, what if each of these if functions was its own function / handler so you are operating on the full arrays each time? Then at the end do something like combine all the masks.

`valid_mask = valid_cycles & valid_times & valid_prior_spins`